### PR TITLE
Optim-wip: Add a ton of missing docs

### PR DESCRIPTION
**Pull request by ProGamerGov**

### CODE REVIEW

**Comment by ProGamerGov** [comment]
The other optim PRs should probably be reviewed before this one.

**Comment by ProGamerGov** [comment]
All the tests should pass once https://github.com/pytorch/captum/pull/656 is merged.

File: captum/optim/_param/image/images.py [hunk 1]
```diff
@@ -137,13 +149,29 @@ def __init__(
...
+        """
+
         fy = self.torch_fftfreq(height)[:, None]
         fx = self.torch_fftfreq(width)[: width // 2 + 1]
         return torch.sqrt((fx * fx) + (fy * fy))
 
     def get_fft_funcs(self) -> Tuple[Callable, Callable, Callable]:
-        """Support older versions of PyTorch"""
+        """
+        Support older versions of PyTorch.
```
Comment (NarineK) [inline captum/optim/_param/image/images.py:168]: nit: Do you mind adding more documentation to this method ? It does more than supporting older PyTorch version, doesn't it ?

File: captum/optim/_param/image/images.py [hunk 2]
```diff
@@ -180,12 +208,31 @@ def torch_fftfreq(v: int, d: float = 1.0) -> torch.Tensor:
...
+        """
+
         scaled_spectrum = self.fourier_coeffs * self.spectrum_scale
         output = self.torch_irfft(scaled_spectrum)
         return output.refine_names("B", "C", "H", "W")
 
 
 class PixelImage(ImageParameterization):
+    """
+    Parameterize a simple image tensor.
```
Comment (NarineK) [inline captum/optim/_param/image/images.py:223]: nit: Do you mind adding more description here ? Parametrizes with what ?

File: captum/optim/_param/image/images.py [hunk 3]
```diff
@@ -212,6 +259,17 @@ def forward(self) -> torch.Tensor:
 class LaplacianImage(ImageParameterization):
     """
     TODO: Fix divison by 6 in setup_input when init is not None.
+    Parameterize an image tensor with a laplacian pyramid.
```
Comment (NarineK) [inline captum/optim/_param/image/images.py:262]: Do you mind adding reference doc to laplacian pyramid technique ?

File: captum/optim/_param/image/images.py [hunk 4]
```diff
@@ -212,6 +259,17 @@ def forward(self) -> torch.Tensor:
 class LaplacianImage(ImageParameterization):
```
Comment (NarineK) [inline captum/optim/_param/image/images.py:259]: usually we document all non-underscored methods in the public classes. If we don't want those methods to be public we can underscore them. I see that we have quite some in `images.py`

File: captum/optim/_param/image/transforms.py [hunk 5]
```diff
@@ -15,13 +15,24 @@ class BlendAlpha(nn.Module):
     r"""Blends a 4 channel input parameterization into an RGB image.
 
     You can specify a fixed background, or a random one will be used by default.
+
+    Args:
```
Comment (NarineK) [inline captum/optim/_param/image/transforms.py:19]: This can go under the `__init__`.

File: captum/optim/_param/image/transforms.py [hunk 6]
```diff
@@ -36,6 +47,13 @@ class IgnoreAlpha(nn.Module):
     r"""Ignores a 4th channel"""
 
     def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """
+        Ignore the alpha channel.
+        Arguments:
```
Comment (NarineK) [inline captum/optim/_param/image/transforms.py:52]: I think that there is some spacing that we need to do here and for Return.
See: https://github.com/pytorch/captum/blob/master/captum/attr/_core/integrated_gradients.py#L132

File: captum/optim/_param/image/transforms.py [hunk 7]
```diff
@@ -36,6 +47,13 @@ class IgnoreAlpha(nn.Module):
...
+            x (torch.Tensor): RGBA image tensor.
+        Returns:
+            rgb (torch.Tensor): RGB image tensor without the alpha channel.
+        """
         assert x.dim() == 4
```
Comment (NarineK) [inline captum/optim/_param/image/transforms.py:57]: Since `klt_transform` and `i1i2i3_transform` are public, it would be good to document them.
`__ init__` for `ToRGB ` needs docs as well.

File: captum/optim/_param/image/transforms.py [hunk 8]
```diff
@@ -93,6 +111,16 @@ def __init__(self, transform: Union[str, torch.Tensor] = "klt") -> None:
...
+        """
+        Args:
+            x (torch.tensor):  A CHW or NCHW RGB or RGBA image tensor.
+            inverse (bool, optional):  Whether to recorrelate or decorrelate colors.
+                Default is set to False.
+        Returns:
+            chw (torch.tensor):  A tensor with it's colors recorrelated or
+                decorrelated.
+        """
+
```
Comment (NarineK) [inline captum/optim/_param/image/transforms.py:123]: The `__init__` for CenterCrop + RandomScale (public methods), RandomSpatialJitter (public methods) need docs for args as well.

**Review by NarineK — COMMENTED** [review]
Thank you very much for the documentation PR. Overall looks great!
I did a quick pass and left some comments.
I also see that we have `Losses`  that need to be documented a bit more thoroughly . 
https://github.com/pytorch/captum/blob/optim-wip/captum/optim/_core/loss.py

In general I think that before open sourcing we can go over docs again, make them a bit more thorough and clean up the formatting, but we can do that in a separate PR as well.

File: captum/optim/_core/optimization.py [hunk 9]
```diff
@@ -123,10 +123,9 @@ def optimize(
             optimizer (Optimizer, optional):  An torch.optim.Optimizer used to
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/optimization.py:123]: Yeah, I'll add the documentation for those variables!

File: captum/optim/_core/output_hook.py [hunk 10]
```diff
@@ -74,6 +74,11 @@ def __del__(self) -> None:
...
     """
     Simple module for collecting activations from model targets.
+
+    Args:
+        model (nn.Module):  The reference to PyTorch model instance.
+        targets (nn.module or list of nn.module):  The target layers to
+            collect activations from.
     """
 
     def __init__(self, model: nn.Module, targets: Iterable[nn.Module]) -> None:
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/output_hook.py:84]: I'll move all the init arguments from class descriptions to the class' init functions.

File: captum/optim/_param/image/images.py (hunk 1 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/images.py:168]: It just ensures that the same FFT operations are carried regardless of whether your PyTorch version has the `torch.fft` update.

File: captum/optim/_core/output_hook.py [hunk 11]
```diff
@@ -74,6 +74,11 @@ def __del__(self) -> None:
 class ActivationFetcher:
     """
     Simple module for collecting activations from model targets.
+
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/output_hook.py:77]: I don't see `ModuleReuseException` being called anywhere, so I think that it's safe to remove it.

File: captum/optim/_param/image/transforms.py (hunk 7 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/transforms.py:65]: Okay, I've add some documentation and moved the init args from the main description to the init function.

File: captum/optim/_param/image/transforms.py (hunk 6 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/transforms.py]: I've tried to format all the docs to match the formatting used in the integrated_gradients file you linked to.

File: captum/optim/_param/image/transforms.py (hunk 5 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/transforms.py]: I've moved it!

File: captum/optim/_param/image/images.py (hunk 4 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/images.py:314]: I've added some more documentation and added underscore to a bunch of the public class methods!

File: captum/optim/_param/image/images.py (hunk 2 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/images.py]: I tried to improve the description, let me know if it still needs to be improved!

File: captum/optim/_core/output_hook.py [hunk 12]
```diff
@@ -74,6 +74,11 @@ def __del__(self) -> None:
...
+    Args:
+        model (nn.Module):  The reference to PyTorch model instance.
+        targets (nn.module or list of nn.module):  The target layers to
```
[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/output_hook.py]: Fixed!

File: captum/optim/_core/optimization.py [hunk 13]
```diff
@@ -115,18 +117,26 @@ def optimize(
...
             stop_criteria (StopCriteria, optional):  A function that is called
                         every iteration and returns a bool that determines whether
                         to stop the optimization.
                         See captum.optim.typing.StopCriteria for details.
             optimizer (Optimizer, optional):  An torch.optim.Optimizer used to
                         optimize the input based on the loss function.
+            loss_summarize_fn (Callable, optional): The function to use for summarizing
+                tensor outputs from loss functions.
+                Default: default_loss_summarize
+            lr: (float): If no optimizer is given, then lr is used as the learning rate
```
Comment (NarineK) [inline captum/optim/_core/optimization.py:132]: nit: lr: (float, optional)

File: captum/optim/_param/image/images.py [hunk 14]
```diff
@@ -36,6 +45,20 @@ def __new__(
...
     def open(cls, path: str, scale: float = 255.0, mode: str = "RGB") -> "ImageTensor":
+        """
+        Load an image file from a URL or local filepath directly into an `ImageTensor`.
+
+        Args:
+
+            path (str): A URL or filepath to an image.
+            scale (float): The image scale to use.
+                Default: 255.0
+            mode (str:) The image loading mode to use.
```
Comment (NarineK) [inline captum/optim/_param/image/images.py:56]: scale (float, optional):
mode (str, optional):

File: captum/optim/_param/image/images.py [hunk 15]
```diff
@@ -73,9 +96,31 @@ def __torch_function__(
...
         self, figsize: Optional[Tuple[int, int]] = None, scale: float = 255.0
     ) -> None:
+        """
+        Display an `ImageTensor`.
+
+        Args:
+
+            figsize (Tuple[int, int], optional): height & width to use
+                for displaying the `ImageTensor` figure.
+            scale (float): Value to multiply the `ImageTensor` by so that
```
Comment (NarineK) [inline captum/optim/_param/image/images.py:106]: , optional

File: captum/optim/_param/image/images.py [hunk 16]
```diff
@@ -73,9 +96,31 @@ def __torch_function__(
...
 
     def export(self, filename: str, scale: float = 255.0) -> None:
+        """
+        Save an `ImageTensor` as an image file.
+
+        Args:
+
+            filename (str): The filename to use when saving the `ImageTensor` as an
+                image file.
+            scale (float): Value to multiply the `ImageTensor` by so that
```
Comment (NarineK) [inline captum/optim/_param/image/images.py:120]: (float, optional)

File: captum/optim/_param/image/transforms.py [hunk 17]
```diff
@@ -298,10 +391,25 @@ class ScaleInputRange(nn.Module):
     """
 
     def __init__(self, multiplier: float = 1.0) -> None:
+        """
+        Args:
+
+            multiplier (float):  A float value used to scale the input.
```
Comment (NarineK) [inline captum/optim/_param/image/transforms.py:397]: (float, optional)

**Review by NarineK — APPROVED** [review]
Thank you for addressing the comments. For couple inputs `optional` was missing otherwise looks great.

File: captum/optim/_param/image/images.py (hunk 14 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/images.py]: Fixed!

File: captum/optim/_param/image/images.py (hunk 15 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/images.py]: Fixed!

File: captum/optim/_param/image/images.py (hunk 16 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/images.py]: Fixed!

File: captum/optim/_param/image/transforms.py (hunk 17 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/transforms.py]: Fixed!

File: captum/optim/_param/image/images.py (hunk 3 above)
Comment (ProGamerGov) [reply captum/optim/_param/image/images.py:319]: I'll add the citation in future PR as I'm having difficulty finding it at the moment. The Distill / Lucid team is also on hiatus right now, and thus I can't get answers as quickly and easily as before.

File: captum/optim/_core/optimization.py (hunk 13 above)
Comment (ProGamerGov) [reply captum/optim/_core/optimization.py]: Fixed!

File: captum/optim/_core/optimization.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (NarineK) [reply captum/optim/_core/optimization.py:127]: @ProGamerGov, the documentation for losses we will add in a separate PR ?https://github.com/pytorch/captum/blob/optim-wip/captum/optim/_core/loss.py

[the remark this answers is not in the document]
Comment (ProGamerGov) [reply captum/optim/_core/optimization.py:127]: @NarineK Yes, we'll add the loss documentation in a separate PR!
